### PR TITLE
More `ObligationForest` improvements

### PR DESCRIPTION
**Pull request by nnethercote**
Following on from #64500, these commits alsomake the code both nicer and faster.

r? @nikomatsakis

### CODE REVIEW

**Comment by nnethercote** [comment]
Some local measurements:
```
inflate-check
        avg: -10.3%     min: -14.1%     max: -0.0%
keccak-check
        avg: -7.6%      min: -11.7%     max: -0.0%
cranelift-codegen-check
        avg: -1.3%      min: -2.2%      max: -0.0%
clap-rs-check
        avg: -0.7%      min: -1.7%      max: 0.0%
serde-check
        avg: -0.7%      min: -1.2%      max: -0.0%
deeply-nested-check
        avg: -0.5%      min: -0.8%      max: 0.0%
unify-linearly-check
        avg: -0.3%      min: -0.6%      max: 0.0%
style-servo-check
        avg: -0.4%      min: -0.6%      max: 0.0%
wg-grammar-check
        avg: -0.3%      min: -0.5%      max: -0.0%
```
Let's see how a CI perf run compares.

@bors try

**Comment by Mark-Simulacrum** [comment]
@rust-timer build 2ae267a3eaa7c42e9208595aedc942da259f9b2e

**Comment by nnethercote** [comment]
Perf results on CI are similar to what I saw locally: big wins on `keccak` and `inflate`, smaller wins elsewhere.

I think the lesson from this PR (especially commits 1, 4 and 5) is "don't rely on the compiler to fully optimize higher-level constructs in the hottest code".

File: src/librustc/traits/fulfill.rs [hunk 1]
```diff
@@ -256,15 +256,20 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
         &mut self,
         pending_obligation: &mut Self::Obligation,
     ) -> ProcessResult<Self::Obligation, Self::Error> {
-        // if we were stalled on some unresolved variables, first check
+        // If we were stalled on some unresolved variables, first check
         // whether any of them have been resolved; if not, don't bother
         // doing more work yet
         if !pending_obligation.stalled_on.is_empty() {
-            if pending_obligation.stalled_on.iter().all(|&ty| {
+            let mut changed = false;
```
Comment (nikomatsakis) [inline src/librustc/traits/fulfill.rs]: Maybe add a comment pointing at this PR? This seems like a change that is likely to get reverted by someone cleaning up the code. Something like:

--

// NB: this `for` loop was originally a call to `all`, but rewriting in this lower-level form was found to be a compilation time win.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by nikomatsakis** [comment]
r=me with comment, thanks @nnethercote!

File: src/librustc/traits/fulfill.rs (hunk 1 above)
Comment (nnethercote) [reply src/librustc/traits/fulfill.rs]: I have done this. Though you should definitely check out #64572, where I have optimized `all()` and a few other iterator methods, with apparently extraordinary effects.

(I also have another pending PR in which I change this loop yet again, specializing it for the common cases of 0 or 1 elements. I am having a good time this week hammering on this part of the code!)

**Comment by nnethercote** [comment]
Because this has a perf impact, let's do:
@bors rollup-never

**Comment by Mark-Simulacrum** [comment]
Since we've already run perf there's probably no reason for that? command is rollup=never btw

**Comment by nnethercote** [comment]
My preference is for perf-affecting PRs to land on their own. If you look at perf.rust-lang.org and see an improvement or regression, and find it's a roll-up, it can be a real pain to work out which PR caused the change.

@bors rollup=never

**Pull request merged**